Fixed buckets histogram aggregator

jon-wei opened the pull request:
This PR adds a new aggregator type that provides fixed buckets histograms (divide a range into a number of evenly-sized buckets with fixed boundaries).

The fixed buckets histogram supports the min, max, and quantile post-aggregators.

Some benchmark results for add performance on the fixed histogram:
https://gist.github.com/jon-wei/51162792b52c50f7ed0648034a473d1e

Some benchmark results for serde/merge/quantile performance are available here:
https://gist.github.com/jon-wei/f731c77b8aeeeaa7e434b0d262c6b131

b-slim commented on docs/content/development/extensions-core/approximate-histograms.md:
```diff
@@ -1,174 +0,0 @@
-<!--
```
why this file is renamed ? we can not see any diff of what this PR is adding to the doc....

jon-wei replied:
it's not just approximate histograms now, hence the name change, only the section on fixed histograms is new

b-slim replied:
in the end of the day every thing is an approximate as far i can tell, please revert the change or add a new file for the new work, i want to be able to read what is new in this approach how it is different from others, when can be used, especially that now we have handful of histograms.

jon-wei replied:
Sure, I've reverted the file rename.

leventov commented:
What are the differences with existing histogram aggregator(s)?

b-slim commented:
Also it is hard to make sense of the benchmarks without touching on the accuracy of the estimate, in the end of the day one would expect to see what it the tradeoff between how much accurate is the estimate vs compute/aggregate/merge time.
Also any reason this is not a separate extension i do not see why it is merged together.

leerho commented:
@leventov

I haven't been following this thread or examined very closely the code that is being put into druid but at a high level and from a practical and historical perspective streaming histogram algorithms fall into basically 3 types: fixed-bin, heuristic/empirical and quantile sketch algorithms.

I should point out that for comparable data (like numerics), histogram algorithms and quantile algorithms are closely related.  If you know the quantiles of a distribution you can create a PMF, CDF or histogram from the quantiles.

- Fixed-bin algorithms: This is the simplest, and most primitive approach, but requires the user to know, up-front, a great deal about the data that they want to collect.  The user must make assumptions about the min and max values, and about how the data is likely distributed. The user then decides how many bins to define between the min and the max and assigns specific split-point values for the boundaries of all the bins. This can be very tricky because the resulting accuracy will depend greatly on how the actual data is distributed. Linear, Gaussian, log-normal, and power-law distributions require very different bin spacings. If the input data has gaps (missing values), zeros, spikes or jumps, or exceeds the user's assumed min and max values, the resulting histogram could be useless. Each bin is a simple counter that records how many events fell within the range of that bin.  The code is very simple as well, it just performs a simple search or lookup for each value and increments the appropriate counter. 

This approach is nearly always the fastest but space could be a problem. In the attempt to achieve high resolution of the result, the user may just increase the number and density of the bins. This can lead to a huge amount of space consumed. 100,000 bins would consume way more space than the heuristic and sketch algorithms.  Because the fixed bin approach is so dependent on the actual data it is the least reliable in terms of accuracy and generally should be avoided except in very specific applications where the input data is well understood beforehand.

- Heuristic/empirical algorithms: There many types in this category, but I think the main ones are centroid-based, moment-based, and dynamic-bin-based.  I'm sure there are others, but they all have various dependencies on the input data stream, but not quite as severe as the fixed-bin algorithms.  

They also can be divided into two groups based on whether the input data covers a narrow or wide dynamic range:  If the input data ranges over many orders of magnitude, then log-based approaches should be used, where the log of each item is computed prior to further processing. This is only useful for data that is always greater than zero (although zero itself can be handled with simple substitution). Excellent candidates include response time or latency data, file sizes, city populations, pebble sizes, geographic or astronomical distances, etc. For this kind of data the above algorithms become log-centroid-based, log-moment-based, and dynamic, log-bin-based.  

I don't want to go into depth about this class of algorithms except to point out that none of them can make rank-accuracy guarantees without making some assumptions about the input data stream.  (value-accuracy guarantees are nearly impossible to make without some assumption about the input data) The best that they can do is something like "if the input data is relatively well-behaved and smooth in its distribution, the the accuracy can be pretty good."  

A [recently published paper](https://arxiv.org/abs/1803.01969) on moment-quantiles was able to establish upper bounds on its rank-accuracy, but only for the "L1" measurement of error, which, unfortunately, is not a practical measure of error for most users. (The "L1" error is the area between the true distribution and the mathematically computed distribution based on k moments computed from the input stream. If the two distributions fit each other like a glove for the full range of the distribution everything is wonderful. But there is no guarantee that the actual fit is not like a square peg in a round hole.  So depending on where along the distribution the user makes his query, the error could be great or horrible.  And the user has no idea or warning if this is happening.)

These heuristic/empirical algorithms can be quite small in size and perform quite well in terms of merge speed, but their dependence on the input data is their biggest weakness.

- Sketch algorithms: True sketch algorithms (like the quantiles algorithm in the DataSketches library) are the only algorithms that can claim independence of the input data. In other words, the user does not need to know, beforehand, the min or max of the stream. The input stream can contain, negative values, zeros, gaps, spikes, humps and bumps in its distribution and the values can range over multiple orders of magnitude. The DS-quantiles algorithm (of those benchmarked) is the only one that claim (mathematically) _data independence_, that the rank accuracy guarantees of the sketch will hold with a specified statistical confidence.  The  DS-quantiles sketch defines its error in terms of "L-infinity" error, which applies to each and all quantile (or rank) queries against the sketch anywhere along the distribution, which is what most users would like to be the case.

Well designed sketch algorithms can be small in size and fast, (but not as small and fast as a well designed moments quantiles algorithm, for example).

As for this particular PR, benchmarking accuracy, size and speed of these various sketches can be fraught with problems:

- The tester must make some assumption about the input distributions streamed into each of the algorithms. The attached benchmarks in this PR mention a random uniform distribution and a gaussian distribution -- neither of which are common in most data. They are also very smooth and well behaved. This kind of input data can perform well for all the algorithms including the fixed-bin and Heuristic/empirical algorithms with proper setup.  So it is hardly a test at all as these input distributions will not provide meaningful separation between the test subjects. It is kinda like using a simple arithmetic problem to determine whether a student is ready for an advanced calculus class.

- Performing benchmark accuracy tests can be tricky and it is critical that the tester be very clear about how the algorithm error is being measured: was it relative or absolute error? was the error based on value or rank? How many trials per measurement were used? (It appears that the number of trials was 25 -- it should be about 2500!) How was the error computed from the trials? RMSE, MSE, SE, RSE, average, average-absolute-error, etc.  It is not clear. If multiple trials are performed, what about the variance? It is the error variance that says more about the accuracy properties of an algorithm than just the average. Was the error distribution biased? All of these questions should be documented. The user should not have to dig through the benchmark test code to figure it out.

- What error guarantees, if any, is the algorithm under test able to provide given the chosen configuration? Did the algorithm meet its guarantee? Note that if one algorithm specifies an L1 error guarantee and another specifies an L-infinity guarantee, the tester needs to understand how to discriminate between those guarantees in his/her tests because it is like comparing apples and oranges.

- It is the tester's responsibility, when publishing benchmarks like this to use input data distributions that actually will discriminate between the algorithms and document what those are.  Then the user can make the decision whether those distributions apply to his/her application. I don't feel that that was done in this particular case so I find these benchmarks uninformative and potentially misleading to a user.

***
I don't believe that Druid benefits by having all of these histogram variants, it can be very confusing to users. Having the moments-quantiles algorithm and the DataSketches quantiles sketch should be sufficient. Then document those two thoroughly in how they should be used and when. The fixed bin and other heuristic/empirical algorithms just add confusion, in spite of decades of their misuse and abuse in the analysis of data.

jon-wei commented:
I've removed the comparison benchmarks between the fixed bins and approx/sketch aggs, since they were flawed/insufficient. 

I've kept the benchmarks for just the fixed bins to give an idea of its performance (without regards to accuracy), since its performance would be pretty consistent across various distributions and mainly depends on the number of buckets used.

(PS: The add performance could be roughly twice as fast without locking (needed for realtime indexing), perhaps a way to allow aggregators to know when they're running within a context that requires synchronization would be useful in general, as a subject for another PR)

------------------
@leerho Thanks for the detailed commentary and feedback!

@b-slim 

> Also any reason this is not a separate extension i do not see why it is merged together.

Mainly, the aggregator works with the min/max/quantile/quantiles post-aggs that are provided by the druid-histogram extension, so it's simpler to have them be in the same extension. The name "druid-histogram" also doesn't preclude a fixed bins implementation, so I felt it was reasonable to keep them in the same extension.

@b-slim 

> Also it is hard to make sense of the benchmarks without touching on the accuracy of the estimate, in the end of the day one would expect to see what it the tradeoff between how much accurate is the estimate vs compute/aggregate/merge time.

Re: accuracy and when to use the fixed bins, I think @leerho gave an excellent breakdown:

> Fixed-bin algorithms: This is the simplest, and most primitive approach, but requires the user to know, up-front, a great deal about the data that they want to collect. The user must make assumptions about the min and max values, and about how the data is likely distributed. The user then decides how many bins to define between the min and the max and assigns specific split-point values for the boundaries of all the bins. This can be very tricky because the resulting accuracy will depend greatly on how the actual data is distributed. Linear, Gaussian, log-normal, and power-law distributions require very different bin spacings. If the input data has gaps (missing values), zeros, spikes or jumps, or exceeds the user's assumed min and max values, the resulting histogram could be useless. Each bin is a simple counter that records how many events fell within the range of that bin. The code is very simple as well, it just performs a simple search or lookup for each value and increments the appropriate counter.

> This approach is nearly always the fastest but space could be a problem. In the attempt to achieve high resolution of the result, the user may just increase the number and density of the bins. This can lead to a huge amount of space consumed. 100,000 bins would consume way more space than the heuristic and sketch algorithms. Because the fixed bin approach is so dependent on the actual data it is the least reliable in terms of accuracy and generally should be avoided except in very specific applications where the input data is well understood beforehand.

Fixed bins would be a "special case", for general use, the sketch aggregators are going to be a better choice.

In addition, a major motivation for this PR is to support users who already have fixed bin histograms from elsewhere and want to import those into Druid.

jon-wei commented on docs/content/development/extensions-core/approximate-histograms.md:
```diff
@@ -91,17 +95,138 @@ query.
+
+### Outlier handling modes
+
+The outlier handling mode specifies what should be done with values outside of the histogram's range. There are three supported modes:
```
Added some details on that

jon-wei commented on docs/content/development/extensions-core/approximate-histograms.md:
```diff
@@ -91,17 +95,138 @@ query.
 
-### Approximate Histogram post-aggregators
+#### Full serialization format
+
```
Agree, it would be nice to have that documented

jon-wei commented on docs/content/development/extensions-core/approximate-histograms.md:
```diff
@@ -91,17 +95,138 @@ query.
+
+### Ingesting existing histograms 
+
+It is also possible to ingest existing fixed buckets histograms. The input must be a Base64 string encoding a byte array that contains a serialized histogram object. Both "full" and "sparse" formats can be used.
```
Reordered this

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1017 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class FixedBucketsHistogram
```
Added more javadocs/comments to the class

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1017 @@
+      long lowerCountFromOther = Math.round(otherHistogram.getCumulativeCount(rangeStart, true));
+      long upperCountFromOther = Math.round(otherHistogram.getCumulativeCount(rangeEnd, false));
+
+      switch (outlierHandlingMode) {
```
I moved the outlier handling to an "OutlierHandler" interface

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1017 @@
+    }
+  }
+
+  // Based off PercentileBuckets code from Netflix Spectator: https://github.com/Netflix/spectator
```
I changed this to a full link, it's more informative and the repo can be determined from the link

I came across that code at one point, saw it was pretty much what I needed, and was Apache licensed, so decided to use it as a base, no reason beyond that

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1017 @@
+  public byte[] toBytesFull(boolean withHeader)
+  {
+    int size = getFullStorageSize(numBuckets);
+    if (withHeader) {
```
I reorganized the serde code, it should be more understandable now, but `toBytesFull` has that mode for the buffer aggregator where the serde header can be skipped for efficiency (the original version of this patch wasn't doing that actually, which is fixed now)

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1017 @@
+
+  public void toSparseBytesHelper(ByteBuffer buf, int nonEmptyBuckets)
+  {
+    buf.put(SERIALIZATION_VERSION);
```
Added some helper methods for header writing and size calculation

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramAggregatorFactory.java:
```diff
@@ -0,0 +1,339 @@
+@JsonTypeName(FixedBucketsHistogramAggregator.TYPE_NAME)
+public class FixedBucketsHistogramAggregatorFactory extends AggregatorFactory
+{
+  private static int DEFAULT_NUM_BUCKETS = 10;
```
No specific reason

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramAggregatorFactory.java:
```diff
@@ -0,0 +1,339 @@
+  public FixedBucketsHistogramAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") String fieldName,
+      @JsonProperty("numBuckets") Integer numBuckets,
```
Added `@Nullable`

jon-wei commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramSerde.java:
```diff
@@ -0,0 +1,190 @@
+        FixedBucketsHistogramAggregatorFactory aggregatorFactory = (FixedBucketsHistogramAggregatorFactory) agg;
+
+        if (rawValue == null) {
+          //throw new UnsupportedOperationException("Null not supported yet.");
```
Removed

jon-wei commented on extensions-core/histogram/src/test/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramAggregationTest.java:
```diff
@@ -0,0 +1,171 @@
+  public void testIngestWithNullsToZeroAndQuery() throws Exception
+  {
+    // Nulls are ignored and not replaced with default for SQL compatible null handling.
+    // This is already tested in testIngestWithNullsIgnoredAndQuery()
```
Combined the tests

jon-wei commented on extensions-core/histogram/src/test/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramTest.java:
```diff
@@ -0,0 +1,1509 @@
+      1f, 1.5f, 2f, 2.5f, 3f, 3.5f, 4f, 4.5f, 5f, 5.5f, 6f, 6.5f, 7f, 7.5f, 8f, 8.5f, 9f, 9.5f, 10f
+  };
+
+  static final float[] VALUES7 = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 12, 12, 15, 20, 25, 25, 25};
```
They're just used as another set of data points, but I duplicated them to avoid coupling the two tests together

jon-wei commented on extensions-core/histogram/src/test/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramTest.java:
```diff
@@ -0,0 +1,1509 @@
+
+    Percentile percentile = new Percentile();
+    percentile.setData(doubles);
+    log.debug("MY-P12.5: " + quantiles[0]);
```
Removed these debug blocks

jon-wei commented on docs/_redirects.json:
```diff
@@ -122,5 +122,6 @@
   {"source": "tutorials/tutorial-the-druid-cluster.html", "target": "cluster.html"},
-  {"source": "development/extensions-core/caffeine-cache.html", "target":"../../configuration/caching.html"}
+  {"source": "development/extensions-core/caffeine-cache.html", "target":"../../configuration/caching.html"},
+  {"source": "development/extensions-core/approximate-histograms.html", "target":"histograms.html"}
```
Yeah, thanks, I removed this now

clintropolis commented on docs/content/development/extensions-core/approximate-histograms.md:
```diff
@@ -92,17 +96,140 @@ query.
 |`lowerLimit`/`upperLimit`|Restrict the approximation to the given range. The values outside this range will be aggregated into two centroids. Counts of values outside this range are still maintained. |-INF/+INF|
 
+## Fixed Buckets Histogram
+
```
nit: I think it's worth calling out here that the fixed bucket implementation is recommended mostly for when you know what you are doing and why, and pointing to datasketches for general purpose histogramming needs.

jon-wei replied:
Added a comment on that

gianm commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogram.java:
```diff
@@ -0,0 +1,1364 @@
+   * Serialization header format:
+   *
+   * byte: serialization version, must be 0x01
+   * byte: encoding mode, 0x01 for full, 0x02 for full
```
One of these must be a typo.

jon-wei replied:
fixed, thanks

jon-wei commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -75,6 +75,7 @@
   public static final byte LONG_LAST_CACHE_TYPE_ID = 0x18;
   public static final byte TIMESTAMP_CACHE_TYPE_ID = 0x19;
   public static final byte VARIANCE_CACHE_TYPE_ID = 0x1A;
+  public static final byte FIXED_BUCKET_HIST_CACHE_TYPE_ID = 0x1B;
```
fixed, thanks

jon-wei commented:
I'm going to go ahead and merge this, I think I've addressed the comments and have approval from @clintropolis

leventov commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/FixedBucketsHistogramBufferAggregator.java:
```diff
@@ -0,0 +1,125 @@
+
+import java.nio.ByteBuffer;
+
+public class FixedBucketsHistogramBufferAggregator implements BufferAggregator
```
Seems to prone to #3956.

@jon-wei @gianm @clintropolis you seem to be writing new aggregators frequently, could you fix #3956 so that each separate aggregator doesn't need to roll it's own concurrency?

gianm replied:
I don't feel that we've reached consensus on #3956 about how to approach the problem. Both high level handling (in IncrementalIndex) and low level handling (in aggregators) have been discussed, but I don't feel we've reached a conclusion about how to fix it.

I also don't think #3956 is very urgent to fix for BufferAggregators right now: they're only used by groupBy v1 in `useOffheap` mode, but in groupBy v1, `aggregate` and `get` are not called concurrently. So nothing bad happens. But it would matter more in the future if IncrementalIndex starts using BufferAggregators.

gianm replied:
I chimed in on #3956 with 2¢.
